### PR TITLE
feat: use color mapping for Line layer

### PR DESCRIPTION
**Pull request by ChitlangeSahas**
closes: #697

Use the color map, when provided from the UI, for the Line Layer.

Proof that it works, override color map to be white, from UI:

* UI code overriding the map for testing.

Capture d’écran, le 2021-11-23 à 11 50 48 AM

* Giraffe Line layer and the hover layer respects the color map. 
 
Capture d’écran, le 2021-11-23 à 11 51 30 AM

 
* With multiple lines:

Capture d’écran, le 2021-11-24 à 11 51 24 AM
 
Also notice the points on the hover line cross hair, they respect the color mapping too. 

Capture d’écran, le 2021-11-24 à 11 52 33 AM

Also adds a storybook example.

Capture d’écran, le 2021-11-24 à 1 36 46 PM

### CODE REVIEW

File: giraffe/src/components/LineLayer.tsx [hunk 1]
```diff
@@ -43,6 +43,10 @@ export const LineLayer: FunctionComponent<Props> = props => {
     shadeAboveY: height,
   }
 
+  if (config.colorMapping) {
+    config.colorMappingCallback(config.colorMapping)
+  }
+
```
Comment (ChitlangeSahas) [inline giraffe/src/components/LineLayer.tsx:49]: We could delete this if this isn't how we want to use the callback.

File: giraffe/src/utils/legend/tooltip.ts [hunk 2]
```diff
@@ -95,7 +96,11 @@ export const getPointsTooltipData = (
   const colors = orderDataByValue(
     hoveredRowIndices,
     sortOrder,
-    hoveredRowIndices.map(i => fillScale(groupColData[i]))
+    hoveredRowIndices.map(i =>
+      colorMapping
+        ? colorMapping.mappings[groupColData[i]].color
+        : fillScale(groupColData[i])
+    )
```
Comment (ChitlangeSahas) [inline giraffe/src/utils/legend/tooltip.ts:103]: This is the tooltip part of logic for color mapping.

File: giraffe/src/components/LineLayer.tsx [hunk 3]
```diff
@@ -43,6 +43,10 @@ export const LineLayer: FunctionComponent<Props> = props => {
     shadeAboveY: height,
   }
 
+  if (config.colorMapping) {
```
Comment (hoorayimhelping) [inline giraffe/src/components/LineLayer.tsx:46]:
```suggestion
  if (config.colorMapping && config.colorMappingCallback) {
```

let's not infer that the `colorMappingCallback` was passed in if the `colorMapping` is, let's be explicit that they're both there

**Review by hoorayimhelping — COMMENTED** [review]
nicely done!

I had one request about inferring the existence of a property.

I think this should be a minor version bump in giraffe/giraffe's and giraffe/stories' package.json. In other words, the new version should be `2.21.0`

Let's also let Timmy weigh in before you ship this.

File: giraffe/src/components/LineLayer.tsx (hunk 3 above)
Comment (ChitlangeSahas) [reply giraffe/src/components/LineLayer.tsx:46]: Very true! I'll fix that.

**Review by TCL735 — CHANGES_REQUESTED** [review]
There should be a storybook that demonstrates this. Extending some existing stories to include an option for the color mapping would be enough.

**Comment by TCL735** [comment]
Also, i think this won't work yet for Static Legend.

**Comment by ChitlangeSahas** [comment]
> Also, i think this won't work yet for Static Legend.

@TCL735  correct, I would have to look at static legend's logic just like the tooltip, I assume.

File: stories/src/data/fluxCSV.ts [hunk 4]
```diff
@@ -86,3 +86,901 @@ export const tooltipFalsyValues = `#group,false,false,true,true,true,false,false
...
+#default,mean,,,,,,,,
+,result,table,_start,_stop,_time,_value,_field,_measurement,sensor_id
+,,0,2021-11-24T18:43:49.742147732Z,2021-11-24T19:43:49.742147732Z,2021-11-24T18:44:00Z,0.3786237915662678,co,airSensors,TLM0100
+,,0,2021-11-24T18:43:49.742147732Z,2021-11-24T19:43:49.742147732Z,2021-11-24T18:44:10Z,0.37735360421915626,co,airSensors,TLM0100
+,,0,2021-11-24T18:43:49.742147732Z,2021-11-24T19:43:49.742147732Z,2021-11-24T18:44:20Z,0.38950567433178224,co,airSensors,TLM0100
+,,0,2021-11-24T18:43:49.742147732Z,2021-11-24T19:43:49.742147732Z,2021-11-24T18:44:30Z,0.3906371504590928,co,airSensors,TLM0100
+,,0,2021-11-24T18:43:49.742147732Z,2021-11-24T19:43:49.742147732Z,2021-11-24T18:44:40Z,0.3958325900427829,co,airSensors,TLM0100
+,,0,2021-11-24T18:43:49.742147732Z,2021-11-24T19:43:49.742147732Z,2021-11-24T18:44:50Z,0.41191279552390064,co,airSensors,TLM0100
+,,0,2021-11-24T18:43:49.742147732Z,2021-11-24T19:43:49.742147732Z,2021-11-24T18:45:00Z,0.4180948178942983,co,airSensors,TLM0100
+,,0,2021-11-24T18:43:49.742147732Z,2021-11-24T19:43:49.742147732Z,2021-11-24T18:45:10Z,0.42034671367558174,co,airSensors,TLM0100
```
Comment (ChitlangeSahas) [inline stories/src/data/fluxCSV.ts:101]: A significant diff is from CSV data.

**Review by TCL735 — APPROVED** [review]

**Review by hoorayimhelping — APPROVED** [review]
Nice work!

**Pull request merged**